File: clojure/ch02/ex2.67.clj
(ns sicp.ch02 (:use clojure.test))

(defn make-leaf [symbol weight]
  ['leaf symbol weight])
(defn leaf? [x]
  (= (first x) 'leaf))
(defn symbol-leaf [x] (nth x 1))
(defn weight-leaf [x] (nth x 2))

(defn symbols [tree]
  (if (leaf? tree)
      [symbol-leaf tree]
      (nth tree 2)))

(defn weight [tree]
  (if (leaf? tree)
      (weight-leaf tree)
      (nth tree 3)))
      
(defn make-code-tree [left right]
  [left
   right
   (conj (symbols left) (symbols right))
   (+ (weight left) (weight right))])
        
(defn left-branch [tree] (first tree))
(defn right-branch [tree] (nth tree 1))

(defn choose-branch [bit branch]
    (cond (= bit 0) (left-branch branch)
          (= bit 1) (right-branch branch)
          :else (throw (Exception. "bad bit -- choose-branch" bit))))
          
(defn decode [bits tree]
  (defn decode-1 [bits current-branch]
    (if (empty? bits)
      []
      (let [next-branch (choose-branch (first bits) current-branch)]
        (if (leaf? next-branch)
            (cons (symbol-leaf next-branch)
                  (decode-1 (rest bits) tree))
            (decode-1 (rest bits) next-branch)))))
  (decode-1 bits tree))
  
(def sample-tree
  (make-code-tree 
    (make-leaf 'A 4)
    (make-code-tree
      (make-leaf 'B 2)
      (make-code-tree 
        (make-leaf 'D 1)
        (make-leaf 'C 1)))))
        
(def sample-message [0 1 1 0 0 1 0 1 0 1 1 1 0])

(deftest test-decode
  (is (= (decode sample-message sample-tree) '[A D A B B C A])))
  
(run-tests)

File: scala/ch02/ex2.67.scala
type Weight = Int

// In contrast with the 2.6Nb exercises, we'll make Leaf a case class and not
// a subclass of TreeWithBranches. Also, we eliminated EmptyTree.
sealed abstract class Tree(val symbols: List[Symbol], val weight: Weight)
case class TreeWithBranches(
  leftBranch: Tree, rightBranch: Tree, 
  override val symbols: List[Symbol], 
  override val weight: Weight) 
    extends Tree(symbols, weight)
case class Leaf(symbol: Symbol, override val weight: Weight) 
    extends Tree(List(symbol), weight)

def makeCodeTree (left: Tree, right: Tree) = 
  TreeWithBranches(
    left, 
    right, 
    left.symbols ++ right.symbols, 
    left.weight + right.weight)
              
sealed abstract class Bit(val value: Int)
case object Zero extends Bit(0)
case object One  extends Bit(1)

class DecodeException(message: String) extends RuntimeException(message)

def decode (bits: List[Bit], tree: TreeWithBranches): List[Symbol] = {
  def decode1 (bits: List[Bit], current: TreeWithBranches): List[Symbol] = bits match {
    case Nil => Nil
    case head :: tail => chooseBranch (head, current) match {
      case Leaf(symbol, _) => symbol :: decode1(tail, tree)
      case next:TreeWithBranches => decode1(tail, next)
    }
  }
  decode1 (bits, tree)
}
  
def chooseBranch (bit: Bit, branch: TreeWithBranches) = bit match {
  case Zero => branch.leftBranch
  case One  => branch.rightBranch
}
  
val sampleTree = 
  makeCodeTree(
    Leaf ('A, 4),
    makeCodeTree(
      Leaf ('B, 2),
      makeCodeTree(
        Leaf ('D, 1),
        Leaf ('C, 1))))
        
val sampleMessage =
  List(Zero, One, One, Zero, Zero, One, Zero, 
       One, Zero, One, One, One, Zero)

import org.scalatest._ 
import org.scalatest.matchers._

object huffmanDecodeSpec extends Spec with ShouldMatchers {

  describe ("decode") {
    it ("should return the sequence of letters from a bit encoding") {
      decode (sampleMessage, sampleTree) should equal (
        List('A, 'D, 'A, 'B, 'B, 'C, 'A))
    }
  }
}
huffmanDecodeSpec execute
